components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/cache/AuthenticationContextLoader.java

```diff
@@ -0,0 +1,296 @@
+            if (applicationConfig != null) {
+                OptimizedApplicationConfig optApplicationConfig = new OptimizedApplicationConfig(applicationConfig,
+                        context.getTenantDomain());
+                context.getSequenceConfig().setOptApplicationConfig(optApplicationConfig);
+                context.getSequenceConfig().setApplicationConfig(null);
+            }
+        }
+    }
+
+    private void loadApplicationConfig(AuthenticationContext context)
+            throws SessionDataStorageOptimizationException {
+
+        if (context.getSequenceConfig() != null) {
+            ApplicationConfig applicationConfig = context.getSequenceConfig().getApplicationConfig();
+            OptimizedApplicationConfig optApplicationConfig = context.getSequenceConfig().getOptApplicationConfig();
+            if (applicationConfig == null && optApplicationConfig != null) {
+                ServiceProvider serviceProvider = reconstructServiceProvider(optApplicationConfig,
+                        context.getTenantDomain());
+                if (serviceProvider == null) {
+                    throw new SessionDataStorageOptimizationException(
+                            String.format("Cannot find the Service Provider by the resource ID: %s tenant domain: %s",
+                                    optApplicationConfig.getServiceProviderResourceId(), context.getTenantDomain()));
+                }
+                ApplicationConfig appConfig = new ApplicationConfig(serviceProvider, context.getTenantDomain());
+                appConfig.setMappedSubjectIDSelected(optApplicationConfig.isMappedSubjectIDSelected());
+                appConfig.setClaimMappings(optApplicationConfig.getClaimMappings());
+                appConfig.setRoleMappings(optApplicationConfig.getRoleMappings());
+                appConfig.setMandatoryClaims(optApplicationConfig.getMandatoryClaims());
+                appConfig.setRequestedClaims(optApplicationConfig.getRequestedClaims());
+                context.getSequenceConfig().setApplicationConfig(appConfig);
+            }
+        }
+    }
+
+    private ServiceProvider reconstructServiceProvider(OptimizedApplicationConfig optApplicationConfig,
+                                                       String tenantDomain)
+            throws SessionDataStorageOptimizationException {
+
+        ServiceProvider serviceProvider;
+        ApplicationManagementServiceImpl applicationManager = (ApplicationManagementServiceImpl)
+                FrameworkServiceDataHolder.getInstance().getApplicationManagementService();
+        try {
+            serviceProvider = applicationManager.getApplicationByResourceId(
+                    optApplicationConfig.getServiceProviderResourceId(), tenantDomain);
+            if (serviceProvider == null) {
+                return null;
+            }
+            serviceProvider.getLocalAndOutBoundAuthenticationConfig().setAuthenticationSteps(
```

Comment:
Can `getLocalAndOutBoundAuthenticationConfig` become null and will it lead to an exception? Should it be better to handle the null check?